src/netlink_pm_upstream.c

```diff
@@ -42,60 +43,462 @@
+               || cmd == MPTCP_PM_CMD_ADD_ADDR);
+
+        /*
+          Payload (nested):
+              Local address family
+              Local address
+              Local port (optional)
+              Local address ID (optional)
+              Token (required for user space MPTCP_PM_CMD_ANNOUNCE)
+              Flags (optional)
+              Network inteface index (optional)
+         */
+
+        // Types chosen to match MPTCP genl API.
+        uint16_t const family = mptcpd_get_addr_family(info->addr);
+        uint16_t const port   = mptcpd_get_port_number(info->addr);
+
+        /*
+          The MPTCP_PM_ADDR_FLAG_SIGNAL flag is required when a port
+          is specified.  Make sure it is set.
+        */
+        if (port != 0)
+                info->flags |= MPTCP_PM_ADDR_FLAG_SIGNAL;
+
+        size_t const payload_size =
+                MPTCPD_NLA_ALIGN(family)
+                + MPTCPD_NLA_ALIGN_ADDR(info->addr)
+                + MPTCPD_NLA_ALIGN_OPT(port)
+                + MPTCPD_NLA_ALIGN_OPT(info->id)
+                + MPTCPD_NLA_ALIGN_OPT(info->flags)
+                + MPTCPD_NLA_ALIGN_OPT(info->ifindex)
+                + MPTCPD_NLA_ALIGN_OPT(token);
+
+        struct l_genl_msg *const msg =
+                l_genl_msg_new_sized(cmd, payload_size);
+
+        bool const appended =
+                append_local_addr_attr(msg, info)
+                && (token == 0
+                    || l_genl_msg_append_attr(
+                            msg,
+                            MPTCP_PM_ATTR_TOKEN,
+                            sizeof(token),  // sizeof(uint32_t)
+                            &token));
+
+        if (!appended) {
+                l_genl_msg_unref(msg);
+
+                return ENOMEM;
+        }
+
+        return l_genl_family_send(pm->family,
+                                  msg,
+                                  mptcpd_family_send_callback,
+                                  (void *) cmd_name, /* user data */
+                                  NULL  /* destroy */) == 0;
+}
+
+// --------------------------------------------------------------
+//          User Space Path Manager Related Functions
+// --------------------------------------------------------------
+static int upstream_announce(struct mptcpd_pm *pm,
+                             struct sockaddr const *addr,
+                             mptcpd_aid_t id,
+                             mptcpd_token_t token)
+{
+        /**
+         * @todo Add support for the optional network interface index
+         *       attribute.
+         */
+        struct addr_info info = {
+                .addr     = addr,
+                .id       = id,
+                .flags    = MPTCP_PM_ADDR_FLAG_SIGNAL,
+                // .ifindex  = ...
+        };
+
+        /**
+         * Set up MPTCP listening socket.
+         *
+         * @todo This should be optional.
+         */
+        if (mptcpd_lm_listen(pm->lm, addr) == 0)
```

Comment:
Should we propagate the port returned from `mptcpd_lm_listen()` to the caller of `mptcpd_pm_add_addr()` or should we require the port to be non-zero to avoid having to do that?  This doesn't prevent the user from explicitly calling `mptcpd_lm_listen()` with a zero port.

Reply:
Yes, I think the caller of `mptcpd_pm_add_addr()` needs to know the port number. Seems like that's needed in order to do `mptcpd_pm_remove_addr()` later?